pkg/operator/starter.go

```diff
@@ -130,3 +135,78 @@ func RunOperator(ctx context.Context, controllerConfig *controllercmd.Controller
 
 	return fmt.Errorf("stopped")
 }
+
+// withCustomConfigDeploymentHook executes the asset as a template to fill out the parts required
+// when using a custom config with controller deployment.
+func withCustomConfigDeploymentHook() csidrivercontrollerservicecontroller.DeploymentHookFunc {
+	return func(_ *opv1.OperatorSpec, deployment *appsv1.Deployment) error {
+		switch used, err := isMultiAZDeployment(); {
+		case err != nil:
+			return fmt.Errorf("could not determine if the cloud has multiple availability zones: %w", err)
+		case !used:
+			return nil
+		}
+		for i := range deployment.Spec.Template.Spec.Containers {
+			container := &deployment.Spec.Template.Spec.Containers[i]
+			if container.Name != "csi-driver" {
+				continue
+			}
+			for i, env := range container.Env {
+				if env.Name == "CLOUD_CONFIG" {
+					container.Env[i].Value = customConfigPath
+					break
+				}
+			}
+			return nil
+		}
+		return fmt.Errorf("could not use multiaz config because the csi-driver container is missing from the deployment")
+	}
+}
+
+// withCustomConfigDaemonSetHook executes the asset as a template to fill out the parts required
+// when using a custom config with node controller daemonset.
+func withCustomConfigDaemonSetHook() csidrivernodeservicecontroller.DaemonSetHookFunc {
+	return func(_ *opv1.OperatorSpec, daemonset *appsv1.DaemonSet) error {
+		switch used, err := isMultiAZDeployment(); {
+		case err != nil:
+			return fmt.Errorf("could not determine if the cloud has multiple availability zones: %w", err)
+		case !used:
+			return nil
+		}
+		for i := range daemonset.Spec.Template.Spec.Containers {
+			container := &daemonset.Spec.Template.Spec.Containers[i]
+			if container.Name != "csi-driver" {
+				continue
+			}
+			for i, env := range container.Env {
+				if env.Name == "CLOUD_CONFIG" {
+					container.Env[i].Value = customConfigPath
+					break
+				}
+			}
+			return nil
+		}
+		return fmt.Errorf("could not use multiaz config because the csi-driver container is missing from the deployment")
+	}
+}
+
+// isMultiAZDeployment returns true if the openstack cloud has multiple availability zones
+func isMultiAZDeployment() (bool, error) {
+	ci, err := getCloudInfo()
+	if err != nil {
+		return false, fmt.Errorf("couldn't collect info about cloud availability zones: %w", err)
+	}
+
+	if len(ci.ComputeZones) == 0 {
+		return false, fmt.Errorf("cloud doesn't have compute availability zones: %w", err)
+	}
+	if len(ci.VolumeZones) == 0 {
+		return false, fmt.Errorf("cloud doesn't have volume availability zones: %w", err)
+	}
+
+	if len(ci.ComputeZones) > 1 || len(ci.VolumeZones) > 1 || ci.ComputeZones[0] != ci.VolumeZones[0] {
```

Comment:
It's a limitation, imo... I agree that we just can check that compute and volume zones are not the same, and set ignore-volume-az only in this case.

But it may cause problems when, for instance, there are 2 compute zones (AZ A, AZ B), and 2 volume zones (AZ A, AZ B). Despite the fact that both lists are equal, users can't mount a volume from AZ B to a pod on a node from AZ A.

I'm don't mind to implement this restriction, but it must be documented, and users should be aware about this behavior.
Or we can keep current condition, which covers all potential scenarios .